File: AoC/2023_problems/day13.js
const { count } = require("console");
const fs = require("fs")

const data = fs.readFileSync("AoC/2023_problems/day13.txt", "utf8")


let test_data = `#.##..##.
..#.##.#.
##......#
##......#
..#.##.#.
..##..##.
#.#.##.#.

#...##..#
#....#..#
..##..###
#####.##.
#####.##.
..##..###
#....#..#`


function transpose(grid) {
    return grid[0].map((col, i) => grid.map(row => row[i]));
}

function countDifferences(a, b) {
    let diff = 0
    if (a.length != b.length) {
        return false;
    }
    for (let i = 0; i < a.length; i++) {
        for (let j = 0; j < a[i].length; j++) {
            if (a[i][j] != b[i][j]) {
                diff++;
            }       
        }
    }
    return diff;
}

function findReflection(grid) {
    let p1 = 0;
    let p2 = 0;
    for (let i = 1; i < grid.length; i++) {
        let top = grid.slice(0, i).reverse();
        let bottom = grid.slice(i);
        top = top.slice(0, bottom.length);
        bottom = bottom.slice(0, top.length); 
        if (countDifferences(top, bottom) == 0) {
            p1 += i;
        }
        if (countDifferences(bottom, top) == 1) {
            p2 += i;
        }
    }
    return [p1, p2];
}

function solve(data) {
    const grids = data.split("\n\n");
    let part1 = 0;
    let part2 = 0;
    for (let g of grids) {
        let grid = []
        for (let line of g.split("\n")) {
            grid.push(line.split(""));
        }
        let p = findReflection(grid);
        let t = findReflection(transpose(grid));
        part1 += 100 * p[0] + t[0];
        part2 += 100 * p[1] + t[1];
    }
    return [part1, part2]
}

console.log(solve(data))

File: AoC/2023_problems/day13.py
test_data = """#.##..##.
..#.##.#.
##......#
##......#
..#.##.#.
..##..##.
#.#.##.#.

#...##..#
#....#..#
..##..###
#####.##.
#####.##.
..##..###
#....#..#"""


with open("AoC/2023_problems/day13.txt", "r") as f:
    data = f.read()


def find_reflection(grid):
    part1, part2 = 0, 0
    for i in range(1, len(grid)):
        top = grid[:i][::-1]
        bottom = grid[i:]
        # Limit both to same size
        size = min(len(top), len(bottom))
        top = top[:size]
        bottom = bottom[:size]
        # If no difference then part 1, if 1 difference part 2
        if find_difference(top, bottom) == 0:
            part1 += i
        if find_difference(top[:size], bottom[:size]) == 1:
            part2 += i
    return part1, part2


def find_difference(grid1, grid2):
    diff = 0
    for i in range(len(grid1)):
        row1 = grid1[i]
        row2 = grid2[i]
        for j in range(len(row1)):
            if row1[j] != row2[j]:
                diff += 1
    return diff


def process_grid(grid):
    orig_p1, orig_p2 = find_reflection(grid)
    transpose_p1, transpose_p2 = find_reflection(list(map(list, zip(*grid))))
    part1 = 100*orig_p1 or transpose_p1
    part2 = 100*orig_p2 or transpose_p2
    return part1, part2


def solve(data):
    grids = data.split("\n\n")
    p1, p2 = 0, 0
    for g in grids:
        grid = [list(x) for x in g.split()]
        part1, part2 = process_grid(grid)
        p1 += part1
        p2 += part2
        
    return p1, p2

print(solve(data))
